### ISSUE TITLE
File input filters not being applied... 

### ISSUE THREAD

**Issue by dkowis** [spring-projects/spring-integration#3245]
**Affects Version(s): 5.2.5.RELEASE** \<Spring Integration version>
---
I think I've found a bug, in that file filters aren't behaving the way I expect:

```java
package test.thingy;

import org.springframework.integration.file.FileReadingMessageSource;
import org.springframework.integration.file.filters.AcceptOnceFileListFilter;
import org.springframework.integration.file.filters.ChainFileListFilter;
import org.springframework.integration.file.filters.RegexPatternFileListFilter;

import java.io.File;
import java.nio.file.Files;

public class TestMain {
    public static void main(String[] args) throws Exception {
        var regexFilter = new RegexPatternFileListFilter("MATCHONLYTHIS");

        var compositeFilter = new ChainFileListFilter<File>();
        compositeFilter.addFilters(regexFilter, new AcceptOnceFileListFilter<>());

        System.out.println("WHAT: " + compositeFilter.accept(new File("THIS IS A DERP")));
        System.out.println("WHAT: " + compositeFilter.accept(new File("MATCHONLYTHIS")));

        var tempDir = Files.createTempDirectory("test");

        var what = org.springframework.integration.file.dsl.Files.inboundAdapter(tempDir.toFile())
                .useWatchService(true)
                .nioLocker()
                .filter(compositeFilter)
                .watchEvents(FileReadingMessageSource.WatchEventType.CREATE)
                .get();

        System.out.println("Is it running? " + what.isRunning());
        what.start();
        System.out.println("Is it running? " + what.isRunning());

        var shouldNotMatch = tempDir.resolve("doesNotMatch");

        Files.writeString(shouldNotMatch, "just some content");

        Thread.sleep(1000);
        var message = what.receive();

        System.out.println("Message should be null, we shouldn't have anything picked up: " + message);

        var shouldMatch = tempDir.resolve("MATCHONLYTHIS");
        Files.writeString(shouldMatch, "just some content");

        var message2 = what.receive();

        System.out.println("Message should not be null, we should have added a thing! " + message2);
    }
}
```

So what I'm seeing:
```
WHAT: false
WHAT: true
Is it running? false
Is it running? true
15:39:52.041 [main] DEBUG org.springframework.integration.file.FileReadingMessageSource - Added to queue: [/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test5922309638897338630/doesNotMatch]
Message should be null, we shouldn't have anything picked up: GenericMessage [payload=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test5922309638897338630/doesNotMatch, headers={file_originalFile=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test5922309638897338630/doesNotMatch, id=b564b163-0491-7d01-0fc5-0ec196be0da9, file_name=doesNotMatch, file_relativePath=doesNotMatch, timestamp=1586464792056}]
15:39:52.062 [main] DEBUG org.springframework.integration.file.FileReadingMessageSource - Added to queue: [/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test5922309638897338630/MATCHONLYTHIS]
Message should not be null, we should have added a thing! GenericMessage [payload=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test5922309638897338630/MATCHONLYTHIS, headers={file_originalFile=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test5922309638897338630/MATCHONLYTHIS, id=f8765a56-ea80-fb9c-853b-6980d1caf815, file_name=MATCHONLYTHIS, file_relativePath=MATCHONLYTHIS, timestamp=1586464792062}]
```

What I think I should see:
```
WHAT: false
WHAT: true
Is it running? false
Is it running? true
Message should be null, we shouldn't have anything picked up: null
15:39:52.062 [main] DEBUG org.springframework.integration.file.FileReadingMessageSource - Added to queue: [/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test5922309638897338630/MATCHONLYTHIS]
Message should not be null, we should have added a thing! GenericMessage [payload=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test5922309638897338630/MATCHONLYTHIS, headers={file_originalFile=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test5922309638897338630/MATCHONLYTHIS, id=f8765a56-ea80-fb9c-853b-6980d1caf815, file_name=MATCHONLYTHIS, file_relativePath=MATCHONLYTHIS, timestamp=1586464792062}]
```

Basically, I'm seeing it pick up both files, despite the regex filter definitely not matching. The first `WHAT:` log indicates that the filter chain is doing the right thing. Am I doing something wrong, or did I find an actual bug?

**Comment by artembilan** [comment]
Can you confirm that it is a problem with the latest version - `5.2.5.RELEASE`?

Just want to be sure that don't talk about something what has been fixed in between...

Thank you for udnerstanding

**Comment by dkowis** [comment]
Of course, forced my dependencies on a spring boot 2.2.1 project as such:

```kotlin
    implementation("org.springframework.integration:spring-integration-file:5.2.5.RELEASE");
    implementation("org.springframework.integration:spring-integration-core:5.2.5.RELEASE")
```

And I am still getting the same results:

```
Should NOT accept: false
    Should Accept: true
Is it running? false
Is it running? true
16:27:49.482 [main] DEBUG org.springframework.integration.file.FileReadingMessageSource - Added to queue: [/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test16222233470837897242/doesNotMatch]
Message should be null, we shouldn't have anything picked up: GenericMessage [payload=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test16222233470837897242/doesNotMatch, headers={file_originalFile=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test16222233470837897242/doesNotMatch, id=d3e1bd66-2c2b-e8d5-f016-3cc499a27a69, file_name=doesNotMatch, file_relativePath=doesNotMatch, timestamp=1586467669497}]
16:27:49.503 [main] DEBUG org.springframework.integration.file.FileReadingMessageSource - Added to queue: [/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test16222233470837897242/MATCHONLYTHIS]
Message should not be null, we should have added a thing! GenericMessage [payload=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test16222233470837897242/MATCHONLYTHIS, headers={file_originalFile=/var/folders/08/3gsfgs015pb9n8z0k4pqdlgh0000gp/T/test16222233470837897242/MATCHONLYTHIS, id=26784d45-c781-998c-4460-8595ec289624, file_name=MATCHONLYTHIS, file_relativePath=MATCHONLYTHIS, timestamp=1586467669503}]
```

**Comment by dkowis** [comment]
Also, of note, bumping the spring boot version to 2.2.6 brought in 5.2.5.RELEASE of everything spring, and I still have the same results.

**Comment by artembilan** [comment]
Good. Will take a look tomorrow.

Thank you!

**Comment by garyrussell** [comment]
You are not using Spring to manage these components; you can't just instantiate them in a `main()` method and expect them to work.

There's a bunch of initialization code that Spring automatically calls during initialization.

If you create these objects outside of Spring, you are responsible for all the setup and initialization.

`onInit()` is called from the `afterPropertiesSet()` method on the super class, which is called by Spring because it implements `InitializingBean`. That's where the filter is passed into the scanner.

There may be other setup you are missing too.

Bottom line is it is simplest to use these components as they are designed to run (in a Spring application context).

https://github.com/spring-projects/spring-integration/blob/97702ae7128bc3280b54f29fa47da6a360368e4d/spring-integration-file/src/main/java/org/springframework/integration/file/FileReadingMessageSource.java#L327-L349

**Comment by dkowis** [comment]
So there's no way to test the `FileReadingMessageSource` without loading up everything else we've got wired together? I'd just like to test that guy in isolation. We have a `@Bean` method that creates it. 

These are probably basic questions, but my googling for how to test individual spring components hasn't lead me in the right directions.

**Comment by garyrussell** [comment]
Typically, in a unit test, you have to "play the part of Spring" to properly initialize the "bean". In this case, if you call `afterPropertiesSet()`, the filter will be injected into the scanner, which is actually where it is used.

More complex beans need other setup - e.g. those that implement `ApplicationContextAware`, `ApplicationEventPublisherAware`, ... (and many more) to get full functionality.

In most cases, just calling `afterPropertiesSet()` is enough, but in other cases you might have to inject a mock `BeanFactory` as well as other setups. In those cases, most components will throw an exception, such as "A BeanFactory is required".

But, it is very common to require a call to `afterPropertiesSet()` (after, ...err, all the properties are set) in order to properly set up the component.

**Comment by artembilan** [comment]
I wonder why do you need to test a third-party component in the unit test...
You have not developed it, so it is not your responsibility to test it in isolation.
 It is fully covered by tests in the library you take it from.

What you need in your tests is a flow using components from the framework.

**Comment by dkowis** [comment]
The thought process was to make sure that we configured the component to get the behavior we expect rather than just assert we filled out the configuration correctly. Following that, it's easier to diagnose errors if we're just looking at one part of the flow, rather than the whole flow. If we had an error on there, it's easier to find if we're testing only one component at a time. 

That's the way I was testing the component. Not an exhaustive test of the component, but to make sure it's behaving the way we configured it to behave. Just enough testing to say "We configured it this way, therefore it should emit messages for these kinds of files." "It should not emit messsages for these kinds of files." Keeps the scope of behavior small.

It made the tasks a lot smaller and easier to do TDD with. I'd thought it'd be easy to test the output of the FileMessageSource, in isolation. Seems that may have been a bad approach.

**Comment by artembilan** [comment]
Well, your approach might be ok, but you need to be sure that you provide all the requirements for the component to test its behavior correctly. And it turned out you have missed an essential part: Spring dependency injection and inversion control container. 

To test a flow partially we suggest to use a `MockIntegration`: https://docs.spring.io/spring-integration/docs/5.2.5.RELEASE/reference/html/testing.html#test-context

**Comment by dkowis** [comment]
All right, thanks! I think this helps us get going well enough to try to clean up things and get this tested the right way.
